## 1. Operations in python at byte code level

### To know :  REPL
R -> Read
E -> Exaluate
P -> Print
L -> loop

### What is byte code 
is a representation 
8 bits -> 1 byte and each bit has some meaning

for example 00000001 -> it performs say 1+1

### Opcode - operational code
it represents a particular function which is defined by us

for example in any general purpose processor any manufacturer will be giving a table which 
opcode does what functions

in python mutiple byte code and make a single opcode
    example lets take 0000001 is a series of byte code and let it be made as an opcode 
    and it may take few parameters which takes few other byte code series say addition
    based the table it does what kind process takes place
    
    Python generates series bytecode and python virtual machine runs it
    
    

### in python if we define a function where do we find it in the memory

In [4]:
def fun(x,y):
    return x+y

In [5]:
print(fun)

<function fun at 0x7fec0312f910>


### this gives above function  the address of the of the function and the defintions address is in the below cell

In [6]:
fun.__code__

<code object fun at 0x7febf0c6e810, file "/tmp/ipykernel_4153/37564102.py", line 1>

### it strats with b' ' which represents the byte code of the function

In [10]:
bytecode = fun.__code__.co_code
print(bytecode)

b'|\x00|\x01\x17\x00S\x00'


### lets store this as a list and this is the bytecode repesentation

In [12]:
list_bytecode = list(bytecode)
print(list_bytecode)

[124, 0, 124, 1, 23, 0, 83, 0]


In [13]:
from dis import dis

dis(fun)

  2           0 LOAD_FAST                0 (x)
              2 LOAD_FAST                1 (y)
              4 BINARY_ADD
              6 RETURN_VALUE


#### dis takes the bytecode and converts it to human readable formate

### in python the opcode is limited  two argumetns that is one instruction and one argument
    124 ->LOALD_FAST  0 ->
    124 ->LOAD_FAST   1 ->
    23 ->BINARY_ADD   0 ->
    83 ->RETURN_VALUE 0 ->
    
    instructions      arguments

### these bytecode based on cpython and other distribution may have different implementaion
for more information in google search for dis in cpython

### This gives the filename in which the function fun is defined

In [15]:
fun.__code__.co_filename

'/tmp/ipykernel_4153/37564102.py'

### this gives the total arguments in the function fun we declared

In [16]:
fun.__code__.co_argcount

2

### in python when we call any function some shared memory is allocated

In [17]:
fun.__code__.co_stacksize

2

##### Here shows 2 and lets see why is takes 2 exactly

### lets see how does it work

In [18]:
fun("hello","world")

'helloworld'

#### when the function is called the first operand is pushed to the stack and then pushes the second operand is pushed on to the stack and the defined task is performed and in the end the result is then pushed on to the stack later it is poped from the stack

### Optimization :
lets see a C file

int main(){
    int x = 10, y = 10;
    int sum = x+y;
    return 0;
}

def main():
    x,y = 10,10
    sum = x+y
    return 0
    

In [20]:
def main():
    x = 10
    y = 10
    sum = x+y
    return 0

dis(main)

  2           0 LOAD_CONST               1 (10)
              2 STORE_FAST               0 (x)

  3           4 LOAD_CONST               1 (10)
              6 STORE_FAST               1 (y)

  4           8 LOAD_FAST                0 (x)
             10 LOAD_FAST                1 (y)
             12 BINARY_ADD
             14 STORE_FAST               2 (sum)

  5          16 LOAD_CONST               2 (0)
             18 RETURN_VALUE


In [21]:
main.__code__.co_stacksize

2

### lets see the c program in the assembly language

you may find a lable with name main

with different Optimisation we get different levels such as O0, O1, O2
and in O0 it is performing similar to the above byte code 
and when it performs the O2 it removes all the unnessesary things and recognises that it only to return 0